### ISSUE TITLE
InvalidSequenceTokenException: The given sequenceToken is invalid. The next expected sequenceToken is:

### ISSUE THREAD

**Issue by nish20490** [lazywithclass/winston-cloudwatch#137]
In our project we have 6 microservices running and we using Seneca for them.
We have a centralized logger implemented using winston.
Recentry we tried to add logs to cloudwatch so we have added Winston cloudwatch.
But wheneverI start my development server, each microservice starts which puts a log to cloudwatch.
So during this, each time I get this InvalidSequenceTokenException for atleast 2 of the microservices start logs.
I am not sure why this is happening.
Any help would be appreciated.

**Comment by nish20490** [comment]
Because of this exception, some logs are not reaching to cloudwatch. We are facing same issue on staging as well as on prod.
Is there a way to retry these missing logs?

**Comment by lazywithclass** [comment]
Are you using 6 different streams? If I remember correctly (it's been years since I've used this library in production), you should assign to each different source of logs.

There should be a best practice somewhere in AWS Cloudwatch documentation explaining this.

**Comment by Lisenish** [comment]
@nish20490 @lazywithclass 

Same happens to me, we have just two services (worker and API) on in AWS, each of them has 2 instances.
Initially services used different streams but instances shared it and everything worked fine but after some time I noticed that services stop log anything after ~6-24 since the last deploy. Restart or deploy fixes it for the next ~6-24 hours.

I saw `DataAlreadyAcceptedException` and `InvalidSequenceTokenException` errors in logs and decided that the reason is the shared streams, I reconfigured them so now each instance has it's own steram - but unfortunately it didn't help, I continue to see the exactly same behaviour and the same errors in logs

**Comment by lazywithclass** [comment]
Did the proposed PR solved your issue? If so I will review and test it in my environment later this week.

**Comment by Lisenish** [comment]
@lazywithclass I tested it yesterday and unfortunately it's not :(. 
Maybe I did something wrong in the PR but after I deployed this fix I stopped seeing `DataAlreadyAcceptedException` or `InvalidSequenceTokenException` errors in the logs but they still suddenly stopped appearing after some time.

**Comment by lazywithclass** [comment]
Ok I will have to give this a look, but I am studying for September's exam session, so I cannot guarantee a prompt fix.

**Comment by kubakrzempek** [comment]
Now the lib upon `InvalidSequenceTokenException` will try to get a new sequence token and retry sending the logs. However, if there are many resources racing for the same stream (e.g., 4 hosts running the same app from behind load balancer), it might result in the `InvalidSequenceTokenException` reappearing, and in this case, the error will be thrown and the logs (from that batch) discarded.

How about introducing an option `retryCount` (or something) where we could decide how many times the lib should try to get the correct token (for example I'm okay if it will try let's say 30 times). Would that work? If so, I can submit a PR.

**Comment by symbianm** [comment]
Any updates? Have the same issue.

**Comment by kallekj** [comment]
I'm also experiencing this issue, it occurs when we have multiple tests running in parallel which overloads the stream, or it tries to reuse the same sequence token.

**Comment by sporteman** [comment]
Any news on that? I'm getting the same issue here.

**Comment by sowmenappd** [comment]
Hey guys, has anyone worked on this issue yet? We're getting a lot of `InvalidSequenceTokenException` because even though our services use separate streams, most of our services have several instances, which eventually hits the error because of the same race condition mentioned by @kubakrzempek.

**Comment by Lisenish** [comment]
@sowmenappd In our case, we just migrated to `pino` with the logs emitted just as `stdout` and streamed to CW by AWS. I think it's a more reliable and stable approach, but in case you want to stream them by yourself there's `pino-cloudwatch
` transport.

**Comment by sowmenappd** [comment]
@Lisenish thank you, yes this definitely seems like a far better approach. However, wouldn't the `pino-cloudwatch` method also suffer from the same issue if there are multiple instances of the same service?

**Comment by kubakrzempek** [comment]
I completely agree with you! Streaming the logs is definitely the way to go, especially when it comes to scaling. It just makes everything so much easier to manage and maintain.
